src/Chapter4.hs

```diff
@@ -629,7 +661,9 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM l r = l >>= \l' -> if l' then r else pure False
+-- this one feels right for some reason but fails tests
+-- andM l r = l >>= \l' -> fmap (l' &&) r
```

Comment:
Your current implementation is correct and works properly :+1: The reason for this is that it doesn't evaluate the second monadic action unless necessary. With this commented implementation you always evaluate both monadic actions. So it returns `Nothing` on `andM (Just False) Nothing` but you can return the value earlier :fast_forward:

Reply:
The only concern I see is handling the lists. As far as I know lists are also a monad. With current implementation:
```haskell
andM [True, False] [False, True]
[False,True,False] -- three elements
```

with commented implementation:

```haskell
andM [True, False] [False, True]
[False,True,False,False] -- four elements
```

Reply:
Well, these are different implementations with different semantics. The monadic version is lazier. As you can notice, for lists it stops on the first element from the first list which is `False`.

If you want to implement a non-lazy `and` function that always applies to all arguments, you write it as you did. Or, you can even write it using only `Applicatives`:

```haskell
andA :: Applicative f => f Bool -> f Bool -> f Bool
andA m1 m2 = fmap (&&) m1 <*> m2
```

This is a common pattern, and there is `liftA2` function for this:

```haskell
andA = liftA2 (&&)
```

When using `Applicative`, you can't implement a lazy version. So these are just different implementations with different semantics, and for your real problems you should use the one that works best for you :+1: 

On a side note, lists are rarely used as Monads in real-life. The `Monad` instance for lists has only a few practical applications. Usually, we have "values or errors", so types like `Maybe` and `Either` are more common, and many functions are optimized to work with them better.

Reply:
Thank you for explanation!